disnake/interactions/base.py

```diff
@@ -649,26 +670,60 @@ def is_done(self) -> bool:
         """
         return self._responded
 
-    async def defer(self, *, ephemeral: bool = False, with_message: bool = False) -> None:
+    @overload
+    async def defer(
+        self,
+        *,
+        with_message: Literal[True],
+        ephemeral: bool = ...,
+    ):
+        ...
+
+    @overload
+    async def defer(
+        self,
+        *,
+        ephemeral: bool = ...,
+    ):
+        ...
+
+    @overload
+    async def defer(
+        self,
+        *,
+        with_message: Literal[False],
+    ):
+        ...
+
+    async def defer(
+        self,
+        *,
+        with_message: bool = MISSING,
+        ephemeral: bool = MISSING,
+    ) -> None:
```

Comment:
This is a weird construct. We should comment why it exists (I assume to improve type hinting?). If there's a reason it has to be this way it should be documented as such, otherwise something like this ought be more appropriate:
```
async def defer(self, *, with_message: Literal[True] = MISSING, ephemeral: Literal[False] = MISSING)
```
Although this itself feels too specific. We should permit users to pass in with_message = False, even if it doesn't make the best of sense, because it might come directly from other logic and writing multiple calls of a function just to get different signatures sucks. 

So, with that in mind, why not just:
```
async def defer(self, *, with_message: bool = None, emphemeral: bool = None)
```
Then having automatic behavior on None, as is standard practice?

Reply:
> We should permit users to pass in with_message = False

The problem with this method is that it is dealing with multiple interaction types. with_message cannot be False for Application Command interaction types, and will error on the api side. I'm not sure how to better type this.

Reply:
deferred this to a different pr